src/pre_process/m_model.fpp

```diff
@@ -549,4 +554,697 @@ contains
+
+            if (tri_area > threshold_bary*cell_area) then
+                num_inner_vertices = Ifactor_bary_3D*ceiling(tri_area/cell_area)
+                total_vertices = total_vertices + num_inner_vertices
+            end if
+        end do
+
+        ! Allocate memory for the new boundary vertices array
+        allocate (interpolated_boundary_v(1:total_vertices, 1:3))
+
+        ! Fill the new boundary vertices array with original and interpolated vertices
+        total_vertices = 0
+        do i = 1, num_triangles
+            ! Loop through the 3 edges of each triangle
+            do j = 1, 3
+                ! Get the coordinates of the two vertices of the current edge
+                x1 = model%trs(i)%v(j, 1)
+                y1 = model%trs(i)%v(j, 2)
+                z1 = model%trs(i)%v(j, 3)
+                x2 = model%trs(i)%v(mod(j, 3) + 1, 1)  ! Next vertex in the triangle (cyclic)
+                y2 = model%trs(i)%v(mod(j, 3) + 1, 2)
+                z2 = model%trs(i)%v(mod(j, 3) + 1, 3)
+
+                ! Compute the length of the edge
+                edge_length = dsqrt((x2 - x1)**2 + (y2 - y1)**2 + (z2 - z1)**2)
+
+                ! Determine the number of segments and interpolation step
+                if (edge_length > cell_width) then
+                    num_segments = Ifactor_3D*ceiling(edge_length/cell_width)
+                    del_x = (x2 - x1)/num_segments
+                    del_y = (y2 - y1)/num_segments
+                    del_z = (z2 - z1)/num_segments
+                else
+                    num_segments = 1
+                    del_x = 0d0
+                    del_y = 0d0
+                    del_z = 0d0
+                end if
+
+                ! Add original and interpolated vertices to the output array
+                do k = 0, num_segments - 1
+                    total_vertices = total_vertices + 1
+                    interpolated_boundary_v(total_vertices, 1) = x1 + k*del_x
+                    interpolated_boundary_v(total_vertices, 2) = y1 + k*del_y
+                    interpolated_boundary_v(total_vertices, 3) = z1 + k*del_z
+                end do
+
+                ! Add the last vertex of the edge
+                total_vertices = total_vertices + 1
+                interpolated_boundary_v(total_vertices, 1) = x2
+                interpolated_boundary_v(total_vertices, 2) = y2
+                interpolated_boundary_v(total_vertices, 3) = z2
+            end do
+
+            ! Interpolate verties that are not on edges
+            x1 = model%trs(i)%v(1, 1)
+            y1 = model%trs(i)%v(1, 2)
+            z1 = model%trs(i)%v(1, 3)
+            x2 = model%trs(i)%v(2, 1)
+            y2 = model%trs(i)%v(2, 2)
+            z2 = model%trs(i)%v(2, 3)
+            x3 = model%trs(i)%v(3, 1)
+            y3 = model%trs(i)%v(3, 2)
+            z3 = model%trs(i)%v(3, 3)
+            tri_area = f_tri_area(x1, y1, z1, x2, y2, z2, x3, y3, z3)
```

Comment:
Same comment about `t_triangle` as above. Maybe the area could be added to that derived type too?

Reply:
many scalar quantities could just be a single array...